auth/api/iam/openid4vp.go

```diff
@@ -158,6 +163,133 @@ func (r Wrapper) handleAuthorizeRequestFromHolder(ctx context.Context, verifier
+	}
+	presentationDefinition, err := r.auth.Holder().PresentationDefinition(ctx, presentationDefinitionURI)
+	if err != nil {
+		return r.sendAndHandleDirectPostError(ctx, oauthError(oauth.InvalidPresentationDefinitionURI, fmt.Sprintf("failed to retrieve presentation definition on %s", presentationDefinitionURI)), responseURI)
+	}
+
+	// at this point in the flow it would be possible to ask the user to confirm the credentials to use
+
+	// all params checked, delegate responsibility to the holder
+	vp, submission, err := r.auth.Holder().BuildPresentation(ctx, walletDID, *presentationDefinition, metadata.VPFormats, nonce)
+	if err != nil {
+		if errors.Is(err, oauth2.ErrNoCredentials) {
+			return r.sendAndHandleDirectPostError(ctx, oauthError(oauth.InvalidRequest, "no credentials available"), responseURI)
+		}
+		return r.sendAndHandleDirectPostError(ctx, oauthError(oauth.ServerError, err.Error()), responseURI)
+	}
+
+	// any error here is a server error, might need a fixup to prevent exposing to a user
+	return r.sendAndHandleDirectPost(ctx, *vp, *submission, responseURI)
+}
+
+// sendAndHandleDirectPost sends OpenID4VP direct_post to the verifier. The verifier responds with a redirect to the client (including error fields if needed).
+// If the direct post fails, the user-agent will be redirected back to the client with an error. (Original redirect_uri).
+func (r Wrapper) sendAndHandleDirectPost(ctx context.Context, vp vc.VerifiablePresentation, presentationSubmission pe.PresentationSubmission, verifierResponseURI string) (HandleAuthorizeRequestResponseObject, error) {
+	redirectURI, err := r.auth.Holder().PostAuthorizationResponse(ctx, vp, presentationSubmission, verifierResponseURI)
+	if err != nil {
+		return nil, err
+	}
+
+	return HandleAuthorizeRequest302Response{
+		HandleAuthorizeRequest302ResponseHeaders{
+			Location: redirectURI,
+		},
+	}, nil
+}
+
+// sendAndHandleDirectPostError sends errors from handleAuthorizeRequestFromVerifier as direct_post to the verifier. The verifier responds with a redirect to the client (including error fields).
+// If the direct post fails, the user-agent will be redirected back to the client with an error. (Original redirect_uri).
+// If no redirect_uri is present, the user-agent will be redirected to the error page.
+func (r Wrapper) sendAndHandleDirectPostError(ctx context.Context, auth2Error oauth.OAuth2Error, verifierResponseURI string) (HandleAuthorizeRequestResponseObject, error) {
+	redirectURI, err := r.auth.Holder().PostError(ctx, auth2Error, verifierResponseURI)
+	if err == nil {
+		return HandleAuthorizeRequest302Response{
+			HandleAuthorizeRequest302ResponseHeaders{
+				Location: redirectURI,
+			},
+		}, nil
+	}
+
+	msg := fmt.Sprintf("failed to post error to verifier @ %s", verifierResponseURI)
+	log.Logger().WithError(err).Error(msg)
+
+	if auth2Error.RedirectURI == nil {
```

Comment:
same TODO as for `sendAndHandleDirectPost`, we probably need a generic OAuth2 user-facing error page

Reply:
True. We should really focus on getting the error to a XIS/ECD callback page so it can decide on the error presenting.